### PR TITLE
Timetable

### PR DESCRIPTION
**Pull request by mike-fam**
### Team Monkey pull request checklist (Jungle edition)

-   [x] Remember to format your code! (`npm run prettier && npm run eslint`)
-   [x] Remember to run the tests! (`npm run test`)
-   [x] Don't forget comments and docstrings!
-   [x] Is it a breaking change? Remember to mention that!

### Description
- Calendar component implemented (Woohoo!)
- API endpoint for courses with course activities that simulate UQ classes.
- Calendar not speaking to backend yet because of design choice. Further discussion needed.
🐵🐒🙊🙉🙈

### CODE REVIEW

File: src/client/AppRouter.tsx [hunk 1]
```diff
@@ -22,8 +23,9 @@ export const AppRouter: React.FunctionComponent<{}> = () => {
                     component={ClassroomPageContainer}
                 />
                 <Route path="/room/:roomId" component={PrivateRoomContainer} />
-                <Route path="/login" component={LoginContainer} />
-                <Route path="/register" component={RegisterContainer} />
+                <Route path="/calendar-test" component={Calendar} />
+                ] <Route path="/login" component={LoginContainer} />
```
Comment (aXises) [inline src/client/AppRouter.tsx:27]: Brackets?

File: src/client/timetable/Calendar.tsx [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    const chosenMonthRange: Array<Date> = eachDayOfInterval({
+        start: firstDayOfMonth,
+        end: lastDayOfMonth,
+    });
+    const nextMonthRange: Array<Date> = eachDayOfInterval({
+        start: lastDayOfMonth,
+        end: lastDayShown,
+    }).filter((date) => date.getTime() !== lastDayOfMonth.getTime());
+    const monthHeading: string = format(firstDayOfMonth, "MMMM, yyyy");
+    return (
```
Comment (aXises) [inline src/client/timetable/Calendar.tsx:48]: Consider splitting the heading to its own component as this component is fairly large.

File: src/server/routes/course.ts [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+        }
+    })
+);
+
+router.delete(
+    "/delete/:course_code",
+    asyncHandler<CourseResponseType, { course_code: string }, {}>(
+        async (req, res) => {
+            const course = await Course.findOneAndDelete(req.params);
+            if (course === null) {
```
Comment (aXises) [inline src/server/routes/course.ts:83]: `!course` should suffice, weird that Mongoose uses `null`

File: src/server/routes/course.ts [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+            return;
+        }
+        res.json(
+            course.activities.filter((activity) => {
+                const codeFilter = req.body.code || activity.code;
+                const typeFilter = req.body.type || activity.type;
+                const { chosenYear, chosenMonth } = req.body.filter;
+                let timeFilter = false;
+                // @ts-ignore
+                // TODO: type fix?
```
Comment (aXises) [inline src/server/routes/course.ts:173]: You can do a for each loop e.g `activity.weeks.forEach((weekOn, i) => { ... })` to get the correct types or if you want to follow functional programming scheme you can try the `Array.reduce()` method.

**Review by aXises — CHANGES_REQUESTED** [review]

File: src/client/AppRouter.tsx (hunk 1 above)
Comment (JZ4012) [inline src/client/AppRouter.tsx:27]: Unsure about brackets here, are they necessary?

Comment (aXises) [reply src/client/AppRouter.tsx:27]: Probably a typo, since its valid JSX syntax none of the linters picked it up

Comment (mike-fam) [reply src/client/AppRouter.tsx:27]: The brackets may be accidentally left there when I was resolving merge conflicts

Comment (mike-fam) [reply src/client/AppRouter.tsx:27]: The brackets may be accidentally left there when I was resolving merge conflicts

**Review by JZ4012 — APPROVED** [review]

File: src/server/routes/course.ts (hunk 4 above)
Comment (mike-fam) [reply src/server/routes/course.ts]: Switched to using Array.some()

File: src/server/routes/course.ts (hunk 3 above)
Comment (mike-fam) [reply src/server/routes/course.ts]: Done

File: src/client/timetable/Calendar.tsx (hunk 2 above)
Comment (mike-fam) [reply src/client/timetable/Calendar.tsx:47]: Done. CalendarHeader is now a separate component.

File: src/client/AppRouter.tsx (hunk 1 above)
Comment (mike-fam) [reply src/client/AppRouter.tsx]: Bracket removed

File: src/client/AppRouter.tsx [hunk 5]
```diff
@@ -22,8 +23,9 @@ export const AppRouter: React.FunctionComponent<{}> = () => {
                     component={ClassroomPageContainer}
                 />
                 <Route path="/room/:roomId" component={PrivateRoomContainer} />
+                <Route path="/calendar-test" component={Calendar} />
                 <Route path="/login" component={LoginContainer} />
-                <Route path="/register" component={RegisterContainer} />
+                <Route path="/register" component={RegisterContainer} />]{" "}
```
Comment (aXises) [inline src/client/AppRouter.tsx:28]: Is that bracket intended?

**Review by aXises — DISMISSED** [review]

File: src/client/AppRouter.tsx (hunk 5 above)
Comment (JZ4012) [reply src/client/AppRouter.tsx:28]: I would remove these before merging.

**Review by aXises — APPROVED** [review]

**Review by JZ4012 — APPROVED** [review]

**Comment by aXises** [comment]
Nice job, lgtm!

File: src/client/AppRouter.tsx (hunk 5 above)
Comment (mike-fam) [reply src/client/AppRouter.tsx]: Yeh I just removed them. Sorry idk why they're there. I don't recall doing anything to that line.

**Review by oscardunstan — COMMENTED** [review]
Noice

**Pull request merged**